String.php

```diff
@@ -755,32 +790,59 @@ public function getWidth ( ) {
     public function getDirection ( ) {
 
         if(null === $this->_direction) {
-
-            if(null === $this->_string)
-                $this->_direction = static::LTR;
-            else
-                $this->_direction = static::getCharDirection(
-                    mb_substr($this->_string, 0, 1)
-                );
+            // Default
+            $this->_direction = static::LTR;
+
+            // Check for LRM or RLM/ARM
+            $hasLRM      = 0 !== preg_match('#\x{200e}#u', $this->_string);
+            $hasRLMOrARM = 0 !== preg_match('#\x{200f}|\x{061c}#u', $this->_string);
+
+            if($hasLRM || $hasRLMOrARM) {
+                if($hasLRM && $hasRLMOrARM)
+                    $this->_direction = static::BIDI;
+                elseif($hasLRM)
+                    $this->_direction = static::LTR;
+                else
+                    $this->_direction = static::RTL;
+            }
+            else {
+                $this->_direction = static::getCharDirection(mb_substr($this->_string, 0, 1));
+
+                // Check for RLE, RLO or RLI in LTR context -> BIDI
+                if(    static::LTR === $this->_direction
+                   &&            0 !== preg_match('#\x{202b}|\x{202e}|\x{2067}#u', $this->_string))
+                        $this->_direction = static::BIDI;
+                // Check for LRE, LRO or LRI in RTL context -> BIDI
+                elseif(static::RTL === $this->_direction
+                       &&        0 !== preg_match('#\x{202a}|\x{202d}|\x{2066}#u', $this->_string))
+                        $this->_direction = static::BIDI;
+                // Check every other character
+                else {
+                    foreach ($this as $char) {
+                        if ($this->_direction !== $this->getCharDirection($char)) {
```

Comment:
Oh, this is critical in this case :-/. It's not the expected behavior. The algorithm is good but it needs more references I think. Maybe Intl has some ref' about that?